synthesizer/process/src/stack/mod.rs

```diff
@@ -324,42 +324,22 @@ impl<N: Network> StackProgram<N> for Stack<N> {
         &self.program_address
     }
 
-    /// Returns `true` if the stack contains the external record.
-    #[inline]
-    fn contains_external_record(&self, locator: &Locator<N>) -> bool {
-        // Retrieve the external program.
-        match self.get_external_program(locator.program_id()) {
-            // Return `true` if the external record exists.
-            Ok(external_program) => external_program.contains_record(locator.resource()),
-            // Return `false` otherwise.
-            Err(_) => false,
-        }
-    }
-
     /// Returns the external stack for the given program ID.
     #[inline]
-    fn get_external_stack(&self, program_id: &ProgramID<N>) -> Result<&Arc<Stack<N>>> {
-        // Retrieve the external stack.
-        self.external_stacks.get(program_id).ok_or_else(|| anyhow!("External program '{program_id}' does not exist."))
-    }
-
-    /// Returns the external program for the given program ID.
-    #[inline]
-    fn get_external_program(&self, program_id: &ProgramID<N>) -> Result<&Program<N>> {
-        match self.program.id() == program_id {
-            true => bail!("Attempted to get the main program '{}' as an external program", self.program.id()),
-            // Retrieve the external stack, and return the external program.
-            false => Ok(self.get_external_stack(program_id)?.program()),
-        }
-    }
-
-    /// Returns the external record if the stack contains the external record.
-    #[inline]
-    fn get_external_record(&self, locator: &Locator<N>) -> Result<&RecordType<N>> {
-        // Retrieve the external program.
-        let external_program = self.get_external_program(locator.program_id())?;
-        // Return the external record, if it exists.
-        external_program.get_record(locator.resource())
+    fn get_external_stack(&self, program_id: &ProgramID<N>) -> Result<Arc<Stack<N>>> {
+        // Check that the program ID is imported by the program.
+        ensure!(self.program.contains_import(program_id), "External program '{program_id}' is not imported.");
+        // Upgrade the weak reference to the process-level stack map and retrieve the external stack.
+        let result = self
+            .stacks
+            .upgrade()
+            .ok_or_else(|| anyhow!("Process-level stack map does not exist"))?
+            .read()
+            .get(program_id)
+            .cloned()
+            .ok_or_else(|| anyhow!("External stack for '{program_id}' does not exist"));
+        // Return the external stack.
+        result
```

Comment:
```suggestion
        self
            .stacks
            .upgrade()
            .ok_or_else(|| anyhow!("Process-level stack map does not exist"))?
            .read()
            .get(program_id)
            .cloned()
            .ok_or_else(|| anyhow!("External stack for '{program_id}' does not exist"))
```
nit: it's only a matter of time before `clippy` complains about an immediate return